lib/stdlib/src/erl_tar.erl

```diff
@@ -225,809 +279,1581 @@ month(10) -> "Oct";
+convert_header(Bin, _Reader) when byte_size(Bin) =:= 0 ->
     eof;
-convert_header(_Bin) ->
+convert_header(_Bin, _Reader) ->
     throw({error, eof}).
 
-%% Basic sanity.  Better set the element size to zero here if the type
-%% always is of zero length.
-
-convert_header1(H) when H#tar_header.typeflag =:= symlink, H#tar_header.size =/= 0 ->
-    convert_header1(H#tar_header{size=0});
-convert_header1(H) when H#tar_header.typeflag =:= directory, H#tar_header.size =/= 0 ->
-    convert_header1(H#tar_header{size=0});
-convert_header1(Header) ->
-    Header.
-
-typeflag(Bin) ->
-    [T] = binary_to_list(Bin, ?th_typeflag+1, ?th_typeflag+1),
-    case T of
-	0  -> regular;
-	$0 -> regular;
-	$1 -> link;
-	$2 -> symlink;
-	$3 -> char;
-	$4 -> block;
-	$5 -> directory;
-	$6 -> fifo;
-	$7 -> regular;
-	_  -> unknown
+%% Creates a partially-populated header record based
+%% on the provided file_info record. If the file is
+%% a symlink, then `link` is used as the link target.
+%% If the file is a directory, a slash is appended to the name.
+fileinfo_to_header(Name, #file_info{}=Fi, Link) when is_list(Name) ->
+    BaseHeader = #tar_header{name=Name,
+                             mtime=Fi#file_info.mtime,
+                             atime=Fi#file_info.atime,
+                             ctime=Fi#file_info.ctime,
+                             mode=Fi#file_info.mode,
+                             uid=Fi#file_info.uid,
+                             gid=Fi#file_info.gid,
+                             typeflag=?TYPE_REGULAR},
+    do_fileinfo_to_header(BaseHeader, Fi, Link).
+
+do_fileinfo_to_header(Header, #file_info{size=Size,type=regular}, _Link) ->
+    Header#tar_header{size=Size,typeflag=?TYPE_REGULAR};
+do_fileinfo_to_header(#tar_header{name=Name}=Header,
+                      #file_info{type=directory}, _Link) ->
+    Header#tar_header{name=Name++"/",typeflag=?TYPE_DIR};
+do_fileinfo_to_header(Header, #file_info{type=symlink}, Link) ->
+    Header#tar_header{typeflag=?TYPE_SYMLINK,linkname=Link};
+do_fileinfo_to_header(Header, #file_info{type=device,mode=Mode}=Fi, _Link)
+  when (Mode band ?S_IFMT) =:= ?S_IFCHR ->
+    Header#tar_header{typeflag=?TYPE_CHAR,
+                      devmajor=Fi#file_info.major_device,
+                      devminor=Fi#file_info.minor_device};
+do_fileinfo_to_header(Header, #file_info{type=device,mode=Mode}=Fi, _Link)
+  when (Mode band ?S_IFMT) =:= ?S_IFBLK ->
+    Header#tar_header{typeflag=?TYPE_BLOCK,
+                      devmajor=Fi#file_info.major_device,
+                      devminor=Fi#file_info.minor_device};
+do_fileinfo_to_header(Header, #file_info{type=other,mode=Mode}, _Link)
+  when (Mode band ?S_IFMT) =:= ?S_FIFO ->
+    Header#tar_header{typeflag=?TYPE_FIFO};
+do_fileinfo_to_header(Header, Fi, _Link) ->
+    {error, {invalid_file_type, Header#tar_header.name, Fi}}.
+
+is_ascii(Str) when is_list(Str) ->
+    not lists:any(fun (Char) -> Char >= 16#80 end, Str);
+is_ascii(<<>>) ->
```

Comment:
This may seem a little pedantic.

Mixing binary matching with other clauses like this will cause non-optimal binary matching code. (See the Efficiency Guide.) If you break out the binary matching clauses to their own `is_ascii1/1` function the matching optimizations can be applied. (Use the `bin_opt_info` option to verify.)

That is:

```
is_ascii(Str) when is_list(Str) ->
    not lists:any(fun (Char) -> Char >= 16#80 end, Str);
is_ascii(Bin) ->
    is_ascii1(Bin).

is_ascii1(<<>>) ->
    true;
is_ascii1(<<C,_Rest/binary>>) when C >= 16#80 ->
    false;
is_ascii1(<<_, Rest/binary>>) ->
    is_ascii1(Rest).
```